examples/vault-cluster-authentication-iam/startup-script-vault.sh

```diff
@@ -0,0 +1,145 @@
+# Normally, an operator would SSH and unseal the server in each node manually or, ideally,
+# it should be auto unsealed https://www.vaultproject.io/docs/enterprise/auto-unseal/index.html
+# For this quick example specifically, we are just running one vault server and unsealing it like this
+# for simplicity as this example focuses on authentication and not on unsealing.
+# For a more detailed example on auto unsealing, check the vault enterprise example.
+echo "$SERVER_OUTPUT" | head -n 3 | awk '{ print $4; }' | xargs -l /opt/vault/bin/vault operator unseal
+
+# Exports the client token environment variable necessary for running the following vault commands
+export VAULT_TOKEN=$(echo "$SERVER_OUTPUT" | head -n 7 | tail -n 1 | awk '{ print $4; }')
+
+
+# ==========================================================================
+# BEGIN GCP IAM AUTH EXAMPLE
+# ==========================================================================
+# Auth methods must be configured in advance before users or machines can authenticate.
+
+# Enables authentication
+# This is an http request, and sometimes fails, hence we retry
+retry \
+  "/opt/vault/bin/vault auth enable gcp" \
+  "Trying to enable gcp authentication"
+
+# To be able to verify authentication attempts with the help of the Google API,
+# Vault needs to have access to a service account with the necessary roles.
+# In this example runs on a Google Compute Instance, which means that the credentials
+# are provided to Vault automatically. but the following command would be necessary
+# for using the GCP auth method outside of GCP, such as locally, for example.
+#
+# vault write auth/gcp/config credentials=@/path/to/credentials.json
+
+# Creates a policy that allows writing and reading from an "example_" prefix at "secret" backend
+/opt/vault/bin/vault policy write "example-policy" -<<EOF
+path "secret/example_*" {
+  capabilities = ["create", "read"]
+}
+EOF
+
+# Creates an authentication role
+# The Vault Role name, Project ID and GCP service account email are being passed by terraform
+# This example will allow GCP resources using a given service account to authenticate
+# Read more at: https://www.vaultproject.io/docs/auth/gcp.html#iam-login
+/opt/vault/bin/vault write \
+  auth/gcp/role/${example_role_name}\
+  project_id="${project_id}" \
+  type="iam" \
+  policies="example-policy" \
+  bound_service_accounts="${client_service_account_email}"
+
+# ==========================================================================
+# END GCP IAM AUTH EXAMPLE
+# ==========================================================================
+
+# Writes some secret, this secret is being written by terraform for test purposes
+# Please note that normally we would never pass a secret this way as it is not secure
+# This is just so we can have a test verifying that our web client instance is
+# authenticating correctly
```

Comment:
I'll repeat: the comments here are super helpful to understanding how this code works! 👍